### ISSUE TITLE
Please consider adding scipy.special.iv to jax.special

### ISSUE THREAD

**Issue by NeilGirdhar** [google/jax#2466]
Scipy implements it [here](https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.iv.html).

This function is needed to implement the [von Mises-Fisher distribution](https://en.wikipedia.org/wiki/Von_Mises%E2%80%93Fisher_distribution), which represents directions on a sphere.  I have a JAX implementation of that [here](https://github.com/NeilGirdhar/efax/blob/master/efax/von_mises.py), but it cannot be used until this function is jittable.

**Comment by mattjj** [comment]
cc @srvasude because he's looked at special functions recently, and is part of the TFP team working on JAX.

**Comment by LoicRaillon** [comment]
Is there any update on this issue? What is the plan here, write a jittable version of the Scipy implementation or a different one? In case it could be useful, Scipy has the function [ivp](https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.ivp.html#scipy.special.ivp) which compute nth derivative of modified Bessel function Iv(z) with respect to z.

**Comment by srvasude** [comment]
FWIW, in TensorFlow Probability we have `tfp.math.bessel_ive` and `tfp.math.bessel_kve` with gradients available for the second parameter (and not the first). These are the exponentially modified Bessel functions, and `tfp.distributions.VonMisesFisher` makes use of them.

**Comment by NeilGirdhar** [comment]
So is this just a matter of porting https://github.com/tensorflow/probability/blob/master/tensorflow_probability/python/math/bessel.py#L778 over to https://github.com/google/jax/blob/master/jax/_src/scipy/special.py?  If so, I could probably do that at some point.

**Comment by srvasude** [comment]
TFP also has a JAX backend: https://www.tensorflow.org/probability/examples/TensorFlow_Probability_on_JAX

That way you can just import the distribution / function and use them in JAX.

**Comment by NeilGirdhar** [comment]
That's awesome!  What's the future plan for JAX?  Is it still desirable to have JAX provide these functions?  Is there any benefit?  Does JAX get the custom gradients from tensorflow-probability?

**Comment by srvasude** [comment]
Re: custom gradients. We have a `@tfp_custom_gradient` wrapper that will do the right thing in TF and JAX. So if you use these through our jax backend, these will also have the right gradients.

**Comment by NeilGirdhar** [comment]
@srvasude That's awesome!  Unfortunately, there seems to be a bug with the gradients (https://github.com/tensorflow/probability/issues/1247).

**Comment by mattjj** [comment]
> What's the future plan for JAX? Is it still desirable to have JAX provide these functions? Is there any benefit? Does JAX get the custom gradients from tensorflow-probability?

Things might have to be decided on a case-by-case basis, but in general using TFP with JAX is a great solution. Folks looking to do statistical/probabilistic computation should probably use it, and not expect everything to be upstreamed into JAX core (which could mean redundant maintenance efforts, less orthogonality, etc).

**Comment by NeilGirdhar** [comment]
The VJP of `tfp.math.bessel_ive` appears to be fixed on nightly, so I'm closing this.

**Comment by NeilGirdhar** [comment]
With the help of AI, I was able to produce [some simple Jax functions](https://github.com/NeilGirdhar/tjax/blob/f867ecdb3f75b2b16706836abec1be1b8e0cbce5/tjax/_src/math_tools.py#L100) that seem to work correctly.  (They pass my very intense efax tests for various properties of the von Mises Fisher distribution, including entropy, log-normalizer, and maximum likelihood estimation.)

Is there any interest in pulling these in Jax?

**Comment by jakevdp** [comment]
My only hesitation here: in the past we've accepted contributions of special functions that have later turned out to not have sufficient accuracy in certain domains (e.g. `lpmn`, `sph_harm`). Supporting those domains is not easy – the end result is that we have these semi-deprecated broken special functions that we can't really fix, and that we can't actually remove because too many downstream users rely on them, broken as they are. And the original contributors have moved on, so the core team is stuck with this maintenance burden.

I really want to avoid ending up in that situation again, so if we add JAX-side implementations of something like `iv`, we need to be sure it's robust enough, and that kind of testing takes a lot of effort.

**Comment by NeilGirdhar** [comment]
Those are fair points!  When you say certain domains, do you mean different types (like float16, float32, complex128), or different part of the domain (like near 0)?   Because the two functions both have a number of terms parameter (200 in both cases) that can be controlled.  You could set it per type, for example.  I didn't look into setting it appropriately.

**Comment by egordon** [comment]
Just want to bump this again as it is also required (up to maybe i10) for the random mixture (non rejection sampling) method for von Mises sampling described in [Abrue 2008](https://doi.org/10.1109%2FTCOMM.2008.060510).
